src/Front/ApplePay.php

```diff
@@ -435,4 +436,25 @@ public function applepay_cancel_order()
 
 	}
 
+	public function is_shipping_required() {
+		$cart = WC()->cart->get_cart();
+
+		$required = true;
+		foreach ( $cart as $cart_item ) {
+			if ( ! empty( $cart_item['product_id'] ) ) {
+				$product = wc_get_product( $cart_item['product_id'] );
+
+				//not required if it enters here
+				if ( ! $product->is_virtual() && ! $product->is_downloadable() ) {
+					return true;
+				}
+
+				$required = false;
+			}
+		}
```

Comment:
This is not very clear to me....
it can be simplified to:
```suggestion
		$required = false;
		foreach ( $cart as $cart_item ) {
			if ( ! empty( $cart_item['product_id'] ) ) {
				$product = wc_get_product( $cart_item['product_id'] );
				//not required if it enters here
				if ( ! $product->is_virtual() && ! $product->is_downloadable() ) {
					return true;
				}
			}
		}
```